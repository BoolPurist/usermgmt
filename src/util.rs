pub mod io_util {
    use log::debug;
    use std::collections::HashSet;
    use std::io;

    pub fn user_input() -> String {
        let mut input = String::new();

        io::stdin()
            .read_line(&mut input)
            .expect("Failed to read user input");

        input = input.trim().to_string();
        input
    }

    pub fn hashset_from_vec_str(data: &[String]) -> HashSet<&str> {
        // HashSet::from_iter(data.iter().cloned())
        data.iter().map(|s| s.as_str()).collect::<HashSet<&str>>()
    }

    pub fn get_new_uid(uids: Vec<i32>, group: crate::Group) -> Option<i32> {
        let result;
        // students start at 10000, staff at 1000
        if group == crate::Group::Student {
            result = uids.into_iter().filter(|&i| i >= 10000).collect::<Vec<_>>();
        } else {
            result = uids
                .into_iter()
                .filter(|&i| (1000..10000).contains(&i))
                .collect::<Vec<_>>();
        }

        let max_value = result.iter().max();
        match max_value {
            Some(max) => {
                debug!("Next available uid is: {}", max + 1);
                Some(max + 1)
            }
            None => None,
        }
    }
}
