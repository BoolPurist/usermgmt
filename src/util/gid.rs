use crate::{config::MgmtConfig, Group};
use derive_more::Display;
use getset::CopyGetters;

#[derive(Debug, CopyGetters, Clone, PartialEq, Eq, Display)]
#[display(fmt = "{}", id)]
pub struct ResolvedGid {
    #[getset(get_copy = "pub")]
    gid: i32,
    #[getset(get_copy = "pub")]
    id: Group,
}

impl ResolvedGid {
    pub fn new(group: Group, config: &MgmtConfig) -> Self {
        let gid = match group {
            Group::Staff => config.staff_gid,
            Group::Student => config.student_gid,
            Group::Faculty => config.faculty_gid,
        };
        Self { gid, id: group }
    }
}

#[cfg(test)]
mod testing {
    use crate::{config::MgmtConfig, util::ResolvedGid, Group};

    #[test]
    fn take_gid_from_config() {
        let (staff_gid, student_gid, faculty_gid) = (42, 200, 3001);
        let config = MgmtConfig {
            staff_gid,
            student_gid,
            faculty_gid,
            ..Default::default()
        };

        assert_eq!(student_gid, ResolvedGid::new(Group::Student, &config).gid());
        assert_eq!(staff_gid, ResolvedGid::new(Group::Staff, &config).gid());
        assert_eq!(faculty_gid, ResolvedGid::new(Group::Faculty, &config).gid());
    }
}
