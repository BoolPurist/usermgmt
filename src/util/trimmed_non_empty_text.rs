use std::fmt::Display;

use anyhow::bail;

use crate::prelude::AppError;

/// Contains text which trimmed and is not empty/only white spaces
#[derive(Clone, Debug)]
pub struct TrimmedNonEmptyText(String);

impl TryFrom<&str> for TrimmedNonEmptyText {
    type Error = AppError;

    fn try_from(value: &str) -> Result<Self, Self::Error> {
        let to_validate = value.trim().to_string();

        if to_validate.is_empty() {
            bail!("Must not be empty or only white spaces");
        } else {
            Ok(Self(to_validate))
        }
    }
}

impl TrimmedNonEmptyText {
    pub fn to_lowercase(self) -> Self {
        Self(self.0.to_lowercase())
    }
}

impl Display for TrimmedNonEmptyText {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.0)
    }
}

impl AsRef<str> for TrimmedNonEmptyText {
    fn as_ref(&self) -> &str {
        &self.0
    }
}

impl From<TrimmedNonEmptyText> for String {
    fn from(value: TrimmedNonEmptyText) -> Self {
        value.0
    }
}
