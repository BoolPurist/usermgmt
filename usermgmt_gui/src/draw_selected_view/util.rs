use eframe::egui;

use crate::current_selected_view::SshConnectionState;

pub fn draw_ssh_credentials(ui: &mut egui::Ui, ssh_state: &mut SshConnectionState) {
    ui.label("Ssh credentails");
    ui.horizontal(|ui| {
        ui.label("Username: ");
        if let Some(username) = ssh_state.username.as_mut() {
            ui.text_edit_singleline(username);
        } else {
            let mut new = String::default();
            if ui.text_edit_singleline(&mut new).changed() {
                ssh_state.username = Some(new);
            }
        }
    });
    ui.horizontal(|ui| {
        ui.label("Password: ");
        if let Some(password) = ssh_state.password.as_mut() {
            ui.add(egui::TextEdit::singleline(password).password(true));
        } else {
            let mut new = String::default();
            if ui
                .add(egui::TextEdit::singleline(&mut new).password(true))
                .changed()
            {
                ssh_state.password = Some(new);
            }
        }
    });
}
