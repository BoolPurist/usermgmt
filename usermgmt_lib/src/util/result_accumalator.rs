use crate::prelude::AppResult;
use anyhow::{anyhow, Context};

/// Allows to collect several errors messages which can be later combined into one error variant for
/// error propagation. If no error message is collected, it resolves to an Ok variant.
pub struct ResultAccumalator {
    errs: Vec<String>,
    base_err_msg: String,
}

impl ResultAccumalator {
    pub fn new(error_msg: String) -> Self {
        Self {
            errs: Default::default(),
            base_err_msg: error_msg,
        }
    }

    /// Collects the given error message as the parameter "err_msg" if the parameter "condition" is
    /// false
    pub fn add_err_if_false(&mut self, condition: bool, err_msg: String) {
        if !condition {
            self.errs.push(err_msg)
        }
    }

    /// Collects the given error message as the parameter "err_msg"
    pub fn add_err(&mut self, err_msg: String) {
        self.errs.push(err_msg)
    }
}

impl From<ResultAccumalator> for AppResult {
    fn from(value: ResultAccumalator) -> Self {
        if value.errs.is_empty() {
            return Ok(());
        }

        let all_errs = Err(anyhow!("{}", value.base_err_msg));

        all_errs.context(value.errs.join("\n"))
    }
}

#[cfg(test)]
mod testing {
    use super::*;

    #[test]
    fn resolve_to_ok_for_no_errors() {
        let mut accumalator = ResultAccumalator::new("Should not resolve to an error.".to_owned());
        accumalator.add_err_if_false(true, "...".to_owned());
        let result = AppResult::from(accumalator);
        assert!(result.is_ok());
    }

    #[test]
    fn resolve_to_errors() {
        let mut accumalator = ResultAccumalator::new("Should not resolve to an error.".to_owned());
        accumalator.add_err_if_false(false, "false".to_owned());
        accumalator.add_err_if_false(true, "true".to_owned());
        accumalator.add_err("added".to_owned());
        let result = AppResult::from(accumalator);
        insta::assert_debug_snapshot!(result.err().unwrap());
    }
}
