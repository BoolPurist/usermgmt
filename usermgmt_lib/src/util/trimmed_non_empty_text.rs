use std::borrow::Borrow;

use anyhow::anyhow;

use crate::prelude::AppError;
use derive_more::{AsRef, Display, Into};
/// Contains text which trimmed and is not empty or only white spaces
#[derive(Clone, Debug, Display, AsRef, Into, PartialEq, Eq, Hash)]
pub struct TrimmedNonEmptyText(String);

impl TryFrom<String> for TrimmedNonEmptyText {
    type Error = AppError;

    fn try_from(value: String) -> Result<Self, Self::Error> {
        let to_validate = value.trim().to_string();

        if to_validate.is_empty() {
            Err(anyhow!("Must not be empty or only white spaces"))
        } else {
            Ok(Self(to_validate))
        }
    }
}

impl Borrow<str> for TrimmedNonEmptyText {
    fn borrow(&self) -> &str {
        self.as_ref()
    }
}

impl TryFrom<&str> for TrimmedNonEmptyText {
    type Error = AppError;

    fn try_from(value: &str) -> Result<Self, Self::Error> {
        value.to_string().try_into()
    }
}

impl TrimmedNonEmptyText {
    pub fn to_lowercase(self) -> Self {
        Self(self.0.to_lowercase())
    }
}

#[cfg(test)]
mod testing {
    use super::*;
    #[test]
    fn error_from_empty_and_only_whitespace() {
        assert!(TrimmedNonEmptyText::try_from("").is_err());
        assert!(TrimmedNonEmptyText::try_from("   ").is_err());
    }
    #[test]
    fn ok_and_trimmed_on_non_empty() {
        assert_eq!(
            TrimmedNonEmptyText::try_from("aaa").unwrap().as_ref(),
            "aaa"
        );
        assert_eq!(
            TrimmedNonEmptyText::try_from("  aaa ").unwrap().as_ref(),
            "aaa"
        );
    }
}
