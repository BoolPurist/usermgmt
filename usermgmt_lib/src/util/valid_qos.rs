use derive_more::{AsRef, Display, Into, IntoIterator};

use crate::{config::MgmtConfig, prelude::AppResult, Group};

use super::TrimmedNonEmptyText;
#[derive(Debug, Display, Into, Clone, AsRef, PartialEq, Eq)]
pub struct ValidQos(String);

impl ValidQos {
    /// # Errors
    ///
    /// - If parameter `value` is not a valid non-empty text. See [`TrimmedNonEmptyText`]
    /// - If parameter `value` is not listed within the given valid quality of services aka parameter `valid_qos`
    pub fn new(value: String, valid_qos: &[impl AsRef<str>]) -> AppResult<Self> {
        let trimmed_not_empty: TrimmedNonEmptyText = value.try_into()?;
        if valid_qos
            .iter()
            .any(|to_check_against| to_check_against.as_ref() == trimmed_not_empty.as_ref())
        {
            Ok(Self(trimmed_not_empty.into()))
        } else {
            Err(anyhow::anyhow!(
                "Given qos {} is none of the valid qoses",
                trimmed_not_empty
            ))
        }
    }

    pub fn default_qos_from_conf(group: Group, config: &MgmtConfig) -> Self {
        let value = match group {
            Group::Staff | Group::Faculty => config.staff_default_qos.clone(),
            Group::Student => config.student_default_qos.clone(),
        };
        Self(value)
    }
}

#[derive(Debug, Clone, IntoIterator)]
#[into_iterator(ref)]
pub struct ValidGroupOfQos(Vec<ValidQos>);

impl ValidGroupOfQos {
    pub fn new(value: Vec<TrimmedNonEmptyText>, valid_qos: &[impl AsRef<str>]) -> AppResult<Self> {
        let content = value
            .into_iter()
            .map(|from| ValidQos::new(from.into(), valid_qos))
            .collect::<AppResult<_>>()?;

        Ok(Self(content))
    }

    pub fn from_group(group: Group, config: &MgmtConfig) -> AppResult<Self> {
        let from_config = match group {
            Group::Staff => config.staff_qos.clone(),
            Group::Student => config.student_qos.clone(),
            Group::Faculty => config.staff_qos.clone(),
        }
        .into_iter()
        .map(|to_convert| ValidQos::new(to_convert, &config.valid_qos))
        .collect::<AppResult<_>>()?;

        Ok(Self(from_config))
    }

    pub fn contains_other_qos(&self, other: &ValidQos) -> bool {
        self.0.iter().any(|next| other == next)
    }
}

impl From<ValidGroupOfQos> for Vec<String> {
    fn from(value: ValidGroupOfQos) -> Self {
        value.0.into_iter().map(|qos| qos.into()).collect()
    }
}

#[cfg(test)]
mod testing {
    use super::*;
    #[test]
    fn error_if_not_valid_qos() {
        let valid_qos: Vec<&str> = vec!["basic", "normal"];
        assert!(ValidQos::new("bas".into(), &valid_qos).is_err());
    }
    #[test]
    fn error_if_not_valid_qos_in_any_number_of_qos() {
        let valid_qos: Vec<&str> = vec!["basic", "normal"];
        assert!(ValidGroupOfQos::new(
            vec!["basic".try_into().unwrap(), "basi".try_into().unwrap()],
            &valid_qos
        )
        .is_err());
    }
    #[test]
    fn ok_if_valid_qos() {
        let valid_qos: Vec<&str> = vec!["basic", "normal"];
        assert_eq!(
            ValidQos::new("basic".into(), &valid_qos).unwrap().as_ref(),
            "basic"
        );
    }
    #[test]
    fn ok_if_valid_group_of_qos() {
        let valid_qos: Vec<&str> = vec!["basic", "normal", "init"];
        assert!(ValidGroupOfQos::new(
            vec!["basic".try_into().unwrap(), "normal".try_into().unwrap()],
            &valid_qos
        )
        .is_ok());
    }
}
