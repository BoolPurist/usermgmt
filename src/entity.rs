use crate::util;
use anyhow::Context;
use log::{debug, error, info, warn};
use std::{fs, str::FromStr};

use crate::{
    cli::UserToAdd, config::MgmtConfig, prelude::AppResult, util::TrimmedNonEmptyText, Group,
};

/// Representation of a user entity.
/// It contains all information necessary to add/modify/delete the user.
/// TODO: Add proper encapsulation via getter and setters
pub struct Entity {
    pub username: String,
    pub firstname: String,
    pub lastname: String,
    pub mail: String,
    pub gid: i32,
    pub group: Group,
    pub default_qos: String,
    /// TODO: Add validation if a present publickey is in valid format, OpenSsh
    pub publickey: String,
    pub qos: Vec<String>,
}

impl Entity {
    pub fn new(to_add: &UserToAdd, config: &MgmtConfig) -> AppResult<Self> {
        let staff_qos = trim_and_check_qos_collection(
            &config.staff_qos,
            "qos for staff from field (staff_qos) in config",
        )?;
        let student_qos = trim_and_check_qos_collection(
            &config.student_qos,
            "qos for students from field (student_qos) in config",
        )?;
        let valid_qos = trim_and_check_qos_collection(
            &config.valid_qos,
            "general valid qos from field (valid_qos) in config",
        )?;

        let staff_default_qos = &config.staff_default_qos;
        let student_default_qos = &config.student_default_qos;

        let mut default_qos = &to_add.default_qos;
        let mut group_str = to_add.group.as_ref().to_lowercase();
        let mut qos = trim_and_check_qos_collection(&to_add.qos, "inidivuel qow's for user")?;

        if !util::is_valid_group(&group_str, &config.valid_slurm_groups) {
            warn!(
                "Invalid group specified: {}. Group field will be student",
                group_str
            );
            group_str = "student".to_string();
        }

        let group = Group::from_str(&group_str).unwrap();
        let gid = Entity::map_groupname_to_gid(group.clone(), config).unwrap();

        if qos.is_empty() || !util::is_valid_qos(&qos, &valid_qos) {
            info!(
                "Specified QOS {:?} are either invalid or empty. Using defaults in conf.toml.",
                qos
            );
            match group {
                Group::Staff => qos = staff_qos,
                Group::Student => qos = student_qos,
                Group::Faculty => qos = staff_qos,
            }
        }

        default_qos = {
            let no_default_qos = default_qos.is_empty();
            let no_valid_default_qos = !util::is_valid_qos(&[default_qos.clone()], &valid_qos);
            if no_default_qos {
                warn!(
                    "Specified default QOS {} is empty. Using the value specified in config.",
                    default_qos
                );
            }
            if no_valid_default_qos {
                warn!(
                  "Specified default QOS {} is not one of the valid qos. Using the value specified in config.
                   Valid qos are {:?}", 
                   default_qos,
                   valid_qos
                );
            }

            if no_default_qos && no_valid_default_qos {
                match group {
                    Group::Staff => staff_default_qos,
                    Group::Student => student_default_qos,
                    Group::Faculty => staff_default_qos,
                }
            } else {
                default_qos
            }
        };

        let mut pubkey_from_file = "".to_string();
        if !to_add.publickey.is_empty() {
            debug!("Received PublicKey file path {}", to_add.publickey);
            let pubkey_result = fs::read_to_string(&to_add.publickey);
            match pubkey_result {
                Ok(result) => pubkey_from_file = result,
                Err(e) => error!("Unable to read PublicKey from file! {}", e),
            }
        }

        return Ok(Entity {
            username: to_add.user.as_ref().to_lowercase(),
            firstname: to_add.firstname.clone().into(),
            lastname: to_add.lastname.clone().into(),
            group,
            default_qos: default_qos.to_string(),
            publickey: pubkey_from_file.trim().to_string(),
            qos: qos.to_vec(),
            gid,
            mail: to_add.mail.to_string(),
        });

        fn trim_and_check_qos_collection(
            qos: &[impl AsRef<str>],
            qos_name: &str,
        ) -> AppResult<Vec<String>> {
            qos.iter()
                .map(|to_trim| {
                    TrimmedNonEmptyText::try_from(to_trim.as_ref())
                        .map(String::from)
                        .with_context(|| format!("Invalid field in {}", qos_name))
                })
                .collect::<AppResult<Vec<String>>>()
        }
    }

    /// Convert Group enum into integer gid
    fn map_groupname_to_gid(group: Group, config: &MgmtConfig) -> Result<i32, ()> {
        match group {
            Group::Staff => Ok(config.staff_gid),
            Group::Student => Ok(config.student_gid),
            Group::Faculty => Ok(config.faculty_gid),
        }
    }
}

impl Default for Entity {
    fn default() -> Self {
        Entity {
            username: "".to_string(),
            firstname: "".to_string(),
            lastname: "".to_string(),
            mail: "".to_string(),
            gid: -1,
            group: Group::Student,
            default_qos: "basic".to_string(),
            publickey: "".to_string(),
            qos: vec![],
        }
    }
}
