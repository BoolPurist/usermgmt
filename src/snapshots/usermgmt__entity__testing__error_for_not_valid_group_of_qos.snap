---
source: src/entity.rs
expression: actual
---
Err(
    "Given qos not_valid is none of the valid qoses",
)
